src/wasm/wasm-type.cpp

```diff
@@ -1283,9 +1291,24 @@ bool SubTyper::isSubType(HeapType a, HeapType b) {
   if (b == HeapType::func) {
     return a.isSignature();
   }
+  if (typeSystem == TypeSystem::Nominal) {
+    // Subtyping must be declared in a nominal system, not derived from
+    // structure, so we will not recurse.
+    std::lock_guard<std::mutex> lock(nominalSuperTypesLock);
```

Comment:
Taking a lock on each call to isSubType sounds worrying. Can the subtype relations be stored on the heap types perhaps?

Reply:
Yeah there are a few options here. The sets of supertypes or subtypes could perhaps be stored directly on the HeapTypeInfos or the relation could be cached in thread-local maps. I want to start with the easiest possible thing, though, which is why I just have a single global source of truth with a single course-grained lock and no caching anywhere.

Reply:
I see. But storing the relationships on the HeapTypes isn't "caching" - that would be the single source of truth? To me it feels most natural actually, but maybe I don't understand the entire design here yet.

Do you want to defer this topic for later? I don't feel strongly about when we resolve it.

Reply:
Ohh, I see what you mean! Yes, that is an excellent idea. I think it's worth updating the implementation now.